html-prototype/generate-data.py

```diff
@@ -0,0 +1,200 @@
+            )
+        ]
+    ),
+    "broken": Incident(
+        2,
+        "Service unreachable",
+        [
+            IncidentUpdate(
+                "Identified", 
+                "We have identified the root cause of the problem and are working on a workaround. A final fix will be scheduled for the next maintenance period.",
+                resolves=True
+            ),
+            IncidentUpdate(
+                "Investigating",
+                "We have determined that the service itself is unreachable. The server running it seems fine."
+            ),
+            IncidentUpdate(
+                "Notified",
+                "We have received an automated warning about the service."
+            )
+        ]
+    )
+}
+
+RESOLVING_UPDATES_MAINTENANCE = [
+    IncidentUpdate(
+        "Started",
+        "Started the scheduled maintenance session. Return to operational state is expected in roughly two hours."
+    )
+]
+
+def w(file, text):
+    file.write(text + "\n")
+
+def get_incident(api, day, additional_hours=0):
+    incident = copy.deepcopy(INCIDENTS[day.status])
+    incident.api = api
+    incident.add_date(day.date, additional_hours)
+    return incident
+
+if __name__ == "__main__":
+    end = date.today()
+    start = end - timedelta(days=90)
+    print(f"Generating data for period {start} to {end}")
+    with open("generated.yaml", "w") as f:
+        w(f, PRELUDE)
+        for api in APIS:
+            w(f, "  {")
+            w(f, f"    name: {api},")
+            days_limited = 0
+            days_broken = 0
+            days_maintenance = 0
+            day_data = []
+            for dt in rrule(DAILY, dtstart=start, until=end):
+                day = DayData(dt)
+                day.status = random.choices(DAY_STATUS, DAY_STATUS_WEIGHTS)[0]
+                day.status_text = STATUS_TEXTS[day.status]
+                if day.status == "limited":
+                    days_limited = days_limited + 1
+                if day.status == "broken":
+                    days_broken = days_broken + 1
+                if day.status == "maintenance":
+                    days_maintenance = days_maintenance + 1
+                if day.status != "fine":
+                    day.incidents.append(get_incident(api, day))
+                    if random.randrange(10) > 8:
+                        day.incidents.append(get_incident(api, day, 7))
+                day_data.append(day)
+            days_total = len(day_data)
+            days_fine = days_total - days_limited - days_broken - days_maintenance
+            uptime = (days_total - (days_limited + days_broken + days_maintenance)) / days_total
+            status = day_data[-1].status
+            status_text = STATUS_TEXTS[status]
+            w(f, f"    status: {status},")
+            w(f, f"    status-text: {status_text},")
+            w(f, f"    uptime: {uptime + 0.5:.2f},")
```

Comment:
I guess this is why choosing string concat for yaml generation instead of generation via the yaml library?